src/index.js

```diff
@@ -143,6 +143,22 @@ Promise.prototype.then = function(onFulfilled, onRejected) {
   return prom;
 };
 
+Promise.prototype.finally = function(callback) {
+  var constructor = this.constructor;
+  return this.then(
+    function(value) {
+      return constructor.resolve(callback()).then(function() {
```

Comment:
Based on the spec and the implementation in Chrome, if the the callback is not callable (not a function), it should not throw an error. It should return a new Promise.

```js
if (typeof callback !== 'function') {
  return Promise.resolve(value);
} 
```

Reply:
Ah, nice you catched that ;) That's a change that should be updated in [Promise.prototype.finally polyfill](https://github.com/matthew-andrews/Promise.prototype.finally/blob/master/finally.js) as well. I pretty much just copied that code. When thinking about it, I am also not sure that we need to do constructor.resolve() at all. We're already in a then- callback so we might be fine with just returning the result. Or am I wrong?

Reply:
I was suggesting putting that code before this.then, outside of the “then” to avoid 2 extra promises being created

Reply:
I see. The thing I was thinking about was whether it would be safe to do:

```js
Promise.prototype.finally = function (callback) {        
        if (typeof callback !== 'function') {
            return this.constructor.resolve(value);
        }
        return this.then(value => {
            callback(); // Not handling return value at all
            return value;
        }, err => {
            callback(); // Not handling return value at all
            return this.constructor.reject(err);
        });
}
```
But I see that it should not comply to the spec as it wouldn't await any promise returned by callback(), so let's forget about it...

Reply:
Wait a sec. How could we put the typecheck before this.then() when we don't yet have a value?

Maybe this would do?

```js
Promise.prototype.finally = function(callback) {
  if (typeof callback !== 'function') {
    return this; // Equal to this.then(value => value);
  }
  var constructor = this.constructor;
  return this.then(
    function(value) {
      return constructor.resolve(callback()).then(function() {
        return value;
      });
    },
    function(reason) {
      return constructor.resolve(callback()).then(function() {
        return constructor.reject(reason);
      });
    }
  );
};
```

Reply:
To support ES3, I also put finally in brackets this time.